core-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/RowDenormalizerAggregator.java

```diff
@@ -0,0 +1,133 @@
+  "User is able to specify the list of fields that should be used in the denormalized record, with " +
+  "an option to use an alias for the output field name. " +
+  "For example, " +
+  "'ADDRESS' in the input is mapped to 'addr' in the output schema. The denormalized data is easier to query.")
+public class RowDenormalizerAggregator extends BatchAggregator<String, StructuredRecord, StructuredRecord> {
+
+  private final RowDenormalizerConfig conf;
+  private Map<String, String> outputMappings;
+  private Set<String> outputFields;
+  private Schema outputSchema;
+  private String keyField;
+
+  public RowDenormalizerAggregator(RowDenormalizerConfig conf) {
+    this.conf = conf;
+  }
+
+  @Override
+  public void prepareRun(BatchAggregatorContext context) throws Exception {
+    if (conf.numPartitions != null) {
+      context.setNumPartitions(conf.numPartitions);
+    }
+  }
+
+  @Override
+  public void configurePipeline(PipelineConfigurer pipelineConfigurer) {
+    validateInputFields(pipelineConfigurer.getStageConfigurer().getInputSchema());
+    pipelineConfigurer.getStageConfigurer().setOutputSchema(initializeOutputSchema());
+  }
+
+  @Override
+  public void initialize(BatchRuntimeContext context) throws Exception {
+    outputFields = conf.getOutputFields();
+    outputMappings = conf.getOutputFieldMappings();
+    keyField = conf.getKeyField();
+    outputSchema = initializeOutputSchema();
+  }
+
+  @Override
+  public void groupBy(StructuredRecord record, Emitter<String> emitter) throws Exception {
+    emitter.emit(record.get(keyField).toString());
+  }
+
+  @Override
+  public void aggregate(String groupKey, Iterator<StructuredRecord> iterator,
+                        Emitter<StructuredRecord> emitter) throws Exception {
+    if (!iterator.hasNext()) {
+      return;
+    }
+    StructuredRecord.Builder builder = StructuredRecord.builder(outputSchema);
+    builder.set(conf.getKeyField(), groupKey);
+    while (iterator.hasNext()) {
+      StructuredRecord record = iterator.next();
+      String fieldName = record.get(conf.getFieldName());
+      fieldName = outputMappings.containsKey(fieldName) ? outputMappings.get(fieldName) : fieldName;
+      if (outputFields.contains(fieldName)) {
+        builder.set(fieldName, record.get(conf.getFieldValue()));
+      }
+    }
+    emitter.emit(builder.build());
+  }
+
+  /**
+   * Builds the output schema, using output fields provided by user.
+   */
+  private Schema initializeOutputSchema() {
+    List<Schema.Field> fields = new ArrayList<>();
+    // Input fields coming from source will be of type String; that's why we are building the output schema as String.
+    fields.add(Schema.Field.of(conf.getKeyField(), Schema.of(Schema.Type.STRING)));
+    for (String outputField : conf.getOutputFields()) {
+      fields.add(Schema.Field.of(outputField, Schema.nullableOf(Schema.of(Schema.Type.STRING))));
```

Comment:
which means you probably need to build the schema for every call to aggregate(), based on whatever comes from the iterator